### DISCUSSION TITLE
Separate processes getting read locked for unapparent reason

### DISCUSSION THREAD

**Discussion by radiantone** [sqlalchemy/sqlalchemy#8343] [Usage Questions]
Hi,
  I have two processes running in separate terminals that both issue similar/identical queries to the database in a repeating cycle, like every 10 seconds or so. Recently, a problem has appeared (that wasn't happening before) where they both hang (read locked?) when performing selects from the same table. 

Any tips?

thanks
D

**Comment by zzzeek** [comment]
this would be related to the [transaction isolation](https://en.wikipedia.org/wiki/Isolation_(database_systems)) settings for the database in question.

some links that may be helpful: 

https://www.postgresql.org/docs/current/transaction-iso.html

https://dev.mysql.com/doc/refman/8.0/en/innodb-locking.html

locking and isolation is a complicated topic so to have any more info would need to have lots of specifics.

**Reply by radiantone** [reply]
Thanks. I have been using postgres (in a docker) just fine for a while. It's a bit confusing that two processes can't query because that's the whole use case for the database. I'm not an expert per se, but I have been using sqlalchemy and postgres for a while (not a db admin tho).

**Reply by zzzeek** [reply]
it sounds like you are causing a deadlock of some kind, and yes they are difficult to deal with but there's a lot of reasons they can happen.  PostgreSQL in particular has a lot of aggressive locking.

**Reply by radiantone** [reply]
The queries are running in a context handler that has not exited and thus not committed, so could the problem be the two queries are both within transactions? When I CTRL-C out of one of the processes, the other resumes.

**Reply by radiantone** [reply]
I thought most of this was handled by sqlalchemy behind the scenes tho

**Reply by zzzeek** [reply]
how would SQLAlchemy "handle" that ?

**Reply by radiantone** [reply]
By telling the database what to do, what mode to be in, whether to lock tables or not lock tables, etc. It's all driven by SQL?

**Reply by radiantone** [reply]
This seems to have fixed my problem tho
    _engine = create_engine(CONFIG.get("database", "uri"), isolation_level='READ UNCOMMITTED')

**Reply by CaselIT** [reply]
> By telling the database what to do, what mode to be in, whether to lock tables or not lock tables, etc. It's all driven by SQL?

generally sqlalchemy does what the user tells it to do. Also most locks are implied by the operation being done, so you cannot tell the db not to lock something when an operation is performed.



> This seems to have fixed my problem tho

If read uncommitted is sufficient for your use case then great!

**Reply by zzzeek** [reply]
> By telling the database what to do, what mode to be in, whether to lock tables or not lock tables, etc. It's all driven by SQL?

the database locks the tables automatically based on isolation level, not SQLAlchemy.  these locks are implicit given a particular sequence of statements and the given isolation level.

So that would leave the only way SQLAlchemy could affect this automatically would be to default the isolation level to READ UNCOMMITTED or even AUTOCOMMIT.   But then you don't have ACID.   That would not be a well received choice of default behaivor.
